Media & Text: Replace the deprecated __experimentalImageSizeControl with ResolutionTool

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
ImageSizeControl (__experimentalImageSizeControl) is used in the Media & text block. It was marked as deprecated in [#51545](https://github.com/WordPress/gutenberg/pull/51545/files#diff-5a9a2234d89e4080e031dbc85aafab7b248af84799f65710ad15ca7ca03bd5f1R34-R38).

This PR replaces __experimentalImageSizeControl with the ResolutionTool.

Partial for https://github.com/WordPress/gutenberg/issues/52223

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Deprecated controls should not be used.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Adds a new tools panel and moves the existing block sidebar options to this panel.
Replaces __experimentalImageSizeControl with the ResolutionTool.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

Deactivate Gutenberg.
Add two new media and text blocks with images. One with default options, one with the "Crop image to fill entire column" option enabled and a non default resolution.
Save.

Activate Gutenberg with the PR applied. Refresh the page and confirm that:
-  There are no block validation errors.
- The options use a tools panel.
- All options, including the resolution, works.
- The reset and the "reset all" works.

Save and view the front of the website. Confirm that the blocks display correctly, and the correct image size is loaded.

## Screenshots or screencast <!-- if applicable -->
Before:
<img width="284" alt="Media & text block with the settings panel opened, before the PR" src="https://github.com/WordPress/gutenberg/assets/7422055/60d68f88-8578-46ee-9a16-65a37862a9ad">

After:
<img width="284" alt="Media & text block with the settings panel opened, with the PR, using the tools item panels" src="https://github.com/WordPress/gutenberg/assets/7422055/2bd59705-3287-41b8-8fed-9703105fbeb9">

<img width="559" alt="Media & text block with the options panel opened." src="https://github.com/WordPress/gutenberg/assets/7422055/3e1b8abb-0c0b-4248-afd6-9520548f6ce5">

carolinan committed:
Replace __experimentalImageSizeControl with ResolutionTool

carolinan commented:
@ajlende. Hi.   I am trying to replace the deprecated __experimentalImageSizeControl in the latest posts block.
The resolution tool works well. But do you know if there is a way to use the Dimension tool to only display the width and height input boxes, and not use the scale? I seem to be able to hide the aspect ratio, but not the scale.

If not, I think we need to fully implement the dimension tool with the scale and aspect ratio for the latest posts block.

I am not sure yet how this will work for existing latest posts blocks that already have the 25/50/75/100 % width setting enabled.

CC @t-hamano

carolinan committed:
Merge branch 'trunk' into update/media-text-image-size-control

carolinan committed:
Merge branch 'trunk' into update/media-text-image-size-control

carolinan committed:
Merge branch 'trunk' into update/media-text-image-size-control

carolinan committed:
Fix CS spacing issue after merge conflict

carolinan committed:
Move the Media Width control

Move the media width control and update the "crop image to fill" label to  better match the changes in https://github.com/WordPress/gutenberg/pull/58447.

carolinan committed:
Merge branch 'trunk' into update/media-text-image-size-control

carolinan committed:
CS: Remove the default value from isShownByDefault.

Remove the default value (true) from isShownByDefault in the ToolsPanelItem.

carolinan commented:
I think that the open panel is in the wrong position, I'll try to look into it closer.
<img width="284" alt="media and text block with the options panel opened." src="https://github.com/WordPress/gutenberg/assets/7422055/2dfec5bd-8e83-4ab2-8b64-e032558066e5">

carolinan committed:
Import the constants for the ToolsPanel popover position

Import and apply the constants for the dropdownMenuProps that positions the popover for the options and reset panel.

t-hamano commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -318,6 +365,17 @@ function MediaTextEdit( { attributes, isSelected, setAttributes } ) {
 
 	const blockEditingMode = useBlockEditingMode();
 
+	function resetAll() {
```
Since this function is only used once, how about handling it directly inside the `resetAll` prop?

t-hamano commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -229,81 +234,123 @@ function MediaTextEdit( { attributes, isSelected, setAttributes } ) {
+			resetAll={ resetAll }
+			dropdownMenuProps={ TOOLSPANEL_DROPDOWNMENU_PROPS }
+		>
+			{ mediaUrl && (
```
I think it would be better to remove this conditional statement. Because always rendering this control was intentional in #58447. See [this comment](https://github.com/WordPress/gutenberg/pull/58447#pullrequestreview-1851744275).

t-hamano commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -318,6 +365,17 @@ function MediaTextEdit( { attributes, isSelected, setAttributes } ) {
+			imageFill: false,
+			mediaAlt: undefined,
+			focalPoint: undefined,
+			mediaWidth: undefined,
```
`mediaWidth` must be explicitly given a default value of 50. Otherwise, the block will be broken when resetting all.


https://github.com/WordPress/gutenberg/assets/54422211/4f9ce904-54b3-4286-9729-4ff9cc415c44

t-hamano commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -229,81 +234,123 @@ function MediaTextEdit( { attributes, isSelected, setAttributes } ) {
+				isShownByDefault
+				hasValue={ () => !! isStackedOnMobile }
+				onDeselect={ () =>
+					setAttributes( { isStackedOnMobile: false } )
```
```suggestion
					setAttributes( { isStackedOnMobile: true  } )
```

My understanding is that "reset" means changing to default values. The default value of hoge is true, so how about changing this to true?

carolinan replied:
Thank you, I don't know how I missed that it is enabled by default.

t-hamano commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -229,81 +234,123 @@ function MediaTextEdit( { attributes, isSelected, setAttributes } ) {
+				<ToolsPanelItem
+					label={ __( 'Media width' ) }
+					isShownByDefault
+					hasValue={ () => !! mediaWidth }
```
```suggestion
					hasValue={ () => mediaWidth !== 50 }
```

I think we need to detect that the default value has been changed.

t-hamano commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -229,81 +234,123 @@ function MediaTextEdit( { attributes, isSelected, setAttributes } ) {
-						isStackedOnMobile: ! isStackedOnMobile,
-					} )
+				isShownByDefault
+				hasValue={ () => !! isStackedOnMobile }
```
```suggestion
				hasValue={ () => ! isStackedOnMobile }
```

I think it is necessary to detect that it has been changed from the default value of `true`.

t-hamano commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -318,6 +365,17 @@ function MediaTextEdit( { attributes, isSelected, setAttributes } ) {
+		setAttributes( {
+			isStackedOnMobile: false,
+			imageFill: false,
+			mediaAlt: undefined,
```
This should be updated with an empty string (`''`). Otherwise, I don't think the alt text will be removed when you click "Reset all".

t-hamano reviewed (commented):
Thanks for the PR! I think it's working correctly overall. If you have any questions about comments I left, please feel free to ask 👍

carolinan committed:
Merge branch 'trunk' into update/media-text-image-size-control

carolinan committed:
CS: Correct the spacing in edit.js after the merge conflict.

Indent the ToolsPanelItem for the Focal Point one more tab stop.

carolinan committed:
Remove the condition that wraps the Media width control

The condition that was used to show or hide the Media width control was removed in a previous pull request, but this was missed when trying to solve merge conflicts.

carolinan committed:
Make sure that the media width resets to "50"

Set the default value for mediaWidth in resetAll() to 50.
Make sure that only values other than the default (50) counts as a change to the media with setting.

carolinan committed:
Reverse the Stack on mobile setting

The option is enabled by default, so it should also reset to the enabled (true) state.

carolinan committed:
Merge branch 'trunk' into update/media-text-image-size-control

carolinan commented:
The reason why I have sort of paused the work on this PR is because the Alt is currently not working for the featured image,
https://github.com/WordPress/gutenberg/issues/60023
And I don't want to make changes here that may end up conflicting with any changes needed for that feature.

carolinan committed:
Merge branch 'trunk' into update/media-text-image-size-control

carolinan committed:
Merge branch 'trunk' into update/media-text-image-size-control

richtabor commented:
I'm seeing this here. If we can keep the existing spacing/design then this works well. 


| Trunk  | Proposed |
| ------------- | ------------- |
|![CleanShot 2024-04-30 at 08 46 09](https://github.com/WordPress/gutenberg/assets/1813435/84a25bc4-727f-4bb2-8391-d23a96864c18)|![CleanShot 2024-04-30 at 08 46 33](https://github.com/WordPress/gutenberg/assets/1813435/ea40861b-e3a3-4bd1-97c6-cac65c332579)|

t-hamano commented:
Sorry for the late feedback.

> If we can keep the existing spacing/design then this works well.

I noticed that [this `TextareaControl` component](https://github.com/WordPress/gutenberg/pull/57540/files#diff-799bc423023028d1e733982e9d106233470b262e34737b34c559ab7b14e001d7R363) is not wrapped in a `ToolsPanelItem` component. Wrapping it in a `ToolsPanelItem` component should resolve any spacing issues.

carolinan commented:
Yeah I think something just went wrong trying to solve one of the merge conflicts.

carolinan committed:
Merge branch 'trunk' into update/media-text-image-size-control

carolinan committed:
Remove function resetAll(), add ToolsPanelItem around the alternative text control.

carolinan commented:
I have updated the reset and the missing tools panel item but it is not ready for review yet.
I think I need to disable the resolution option for the featured image.

t-hamano commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -267,86 +272,136 @@ function MediaTextEdit( {
+					/>
+				</ToolsPanelItem>
 			) }
 			{ mediaType === 'image' && (
```
> I think I need to disable the resolution option for the featured image.

Is it possible to solve the problem by changing the conditional statement as follows?

```
mediaType === 'image'  && ! useFeaturedImage && (
```

carolinan replied:
Yeah I just wanted to try and see if I could make it work before disabling it.

t-hamano commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -267,86 +272,136 @@ function MediaTextEdit( {
-					slug={ mediaSizeSlug }
+				<ResolutionTool
+					value={ mediaSizeSlug }
 					imageSizeOptions={ imageSizeOptions }
```
```suggestion
					options={ imageSizeOptions }
```

I think this is the correct prop.

carolinan commented:
Well, I learnt that `__experimentalImageSizeControl` is no longer deprecated: https://github.com/WordPress/gutenberg/pull/56414
Which is the correct component to use?

carolinan committed:
Hide the resolution tool when the featured image is used

t-hamano commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -267,86 +272,136 @@ function MediaTextEdit( {
-		<PanelBody title={ __( 'Settings' ) }>
-			<RangeControl
-				__nextHasNoMarginBottom
-				__next40pxDefaultSize
```
Isn't it better to leave `__next40pxDefaultSize` as is?

carolinan replied:
Yes, this is meant to be untouched! thank you.

t-hamano reviewed (approved):
LGTM!

- ✅ Items whose values have changed from their default values can be reset to their default values individually from the Settings option popover.
- ✅ Can return all values to their default values using the "Reset all" button in the Settings option popover.
- ✅ If "Use featured image" is enabled, the Resolution option will not be displayed.

> Well, I learnt that `__experimentalImageSizeControl` is no longer deprecated: https://github.com/WordPress/gutenberg/pull/56414
> Which is the correct component to use?

I think it's fine as is. The `__experimentalImageSizeControl` component might be deprecated again in the future and a warning may be issued. Instead, it may make more sense to unlock and use the `ResolutionTool` component that is explicitly marked as private.

carolinan committed:
Restore __next40pxDefaultSize on the RangeControl

carolinan committed:
Merge branch 'trunk' into update/media-text-image-size-control
